### DISCUSSION TITLE
Subclassing Array using __new__

### DISCUSSION THREAD

**Discussion by martijnzm** [micropython/micropython#14059] [General]
I am creating a subclass of array.array with some additional functions to use the class for storing measured data.
At first I used `__init__(self, ..)` to create additional members but since `__init__` is called too late in influence the creation of the array, I am now using `__new__(cls)`.
I have to distinguish between cpyhton and micropython for the call to `__new__`. My code:
```python
import array
import platform

def is_micropython():
    try:
        # does not exist under micropython
        platform.python_version()
    except AttributeError:
        return True
    else:
        return False

class MeasurementArray(array.array):
    """Memory efficient storage for a series measurments values.
    Timestamps of the data are not stored, so the data needs to be equidistant in time.
    """

    def __new__(cls, name, typecode="H", length=256, interval=1):
        if is_micropython():
            obj = super().__new__(cls)
        else:
            obj = super().__new__(cls, typecode, [0] * length)
        obj.name = name
        obj.interval = interval
        return obj

    def other_feaature(self):
        print("I have nice features a well")


s = MeasurementArray("temperature", typecode="B", length=21, interval=3)
print("s is a MeasurementArray: " + str(isinstance(s, MeasurementArray)))
print("s is an Array: " + str(isinstance(s, array.array)))
print("type(s): " + str(type(s)))
len(s)  # this works fine on CPython but fails on micropython
s[0] = 1  # this works fine on CPython but fails on micropython
```

Creation of the array works fine under both cpython and micropython. But addressing array elements fails under micropython. The output I get is:
```
s is a MeasurementArray: True
s is an Array: True
type(s): <class 'MeasurementArray'>
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "ArraySubclass.py", line 37, in <module>
TypeError: unsupported type for __len__: 'function'
```
So when calling `len()` micropython does not recognise `s` to be an MeasurementArray. Indexing or calling len() on a standard array does work fine. Since this works as it should under cpython I have been going through the differences between cpython and micropython but I did not find a solution so far.

Why does micropython fail on this code?

**Comment by glenn20** [comment]
Micropython's implementation of `__new__()` is incomplete for classes based on native classes.

In [`micropython/py/objobject.c`](https://github.com/micropython/micropython/blob/d712feb68a3ac8ec181ff75a4ac55997141ad075/py/objobject.c#L50):

```c
static mp_obj_t object___new__(mp_obj_t cls) {
    if (!mp_obj_is_type(cls, &mp_type_type) || !mp_obj_is_instance_type((mp_obj_type_t *)MP_OBJ_TO_PTR(cls))) {
        mp_raise_TypeError(MP_ERROR_TEXT("arg must be user-type"));
    }
    // This executes only "__new__" part of instance creation.
    // TODO: This won't work well for classes with native bases.
    // TODO: This is a hack, should be resolved along the lines of
    // https://github.com/micropython/micropython/issues/606#issuecomment-43685883
    const mp_obj_type_t *native_base;
    return MP_OBJ_FROM_PTR(mp_obj_new_instance(MP_OBJ_TO_PTR(cls), &native_base));
}
```

**Comment by martijnzm** [comment]
OK, thanks for the clarification. Since micropython development is seems to be stalled and circuitpython does not support my ESP8266, I can better implement my own iterator class using a byte array as attribute to store the data.

**Reply by jimmo** [reply]
> Since micropython development is seems to be stalled

Since when?

It's probably pretty unlikely we're going to implement `__new__` for native bases though, it would require quite significant changes and/or code size increase, and hasn't been something that has been requested very often.

**Comment by stinos** [comment]
> I am creating a subclass of array.array with some additional functions to use the class for storing measured data.

Just a note, but from a software architecture POV this doesn't sound like the best idea, and a typcial place where instead of inheritance you'd use composition or extension functions etc.
